File: src/main/java/io/vproxy/vfx/ui/stage/MaxResetButton.java
package io.vproxy.vfx.ui.stage;

import io.vproxy.vfx.theme.Theme;
import javafx.scene.image.Image;
import javafx.scene.layout.CornerRadii;

public class MaxResetButton extends WindowControlButton {
    private final Image[] maxImg;
    private final Image[] rstImg;
    private CornerRadii cornerRadii;

    public MaxResetButton(VStage stage, VStageInitParams initParams) {
        super(stage, initParams);
        maxImg = new Image[]{
            Theme.current().windowMaximizeButtonNormalImage(),
            Theme.current().windowMaximizeButtonHoverImage(),
        };
        rstImg = new Image[]{
            Theme.current().windowResetWindowSizeButtonNormalImage(),
            Theme.current().windowResetWindowSizeButtonHoverImage(),
        };
        updateImage();
    }

    @Override
    protected void init(VStageInitParams initParams) {
        if (initParams.iconifyButton) {
            cornerRadii = CornerRadii.EMPTY;
        } else {
            cornerRadii = new CornerRadii(0, 0, 0, 4, false);
        }
    }

    @Override
    protected void onMouseEntered() {
        super.onMouseEntered();
        imageView.setImage(currentImageGroup()[1]);
    }

    @Override
    protected void onMouseExited() {
        super.onMouseExited();
        imageView.setImage(currentImageGroup()[0]);
    }

    @Override
    protected void onMouseClicked() {
        stage.setMaximized(!stage.isMaximized());
    }

    @Override
    protected CornerRadii getCornerRadii() {
        return cornerRadii;
    }

    @Override
    protected Image getNormalImage() {
        return Theme.current().windowMaximizeButtonNormalImage();
    }

    @Override
    protected Image getHoverImage() {
        return Theme.current().windowMaximizeButtonHoverImage();
    }

    private Image[] currentImageGroup() {
        if (stage.isMaximized()) {
            return rstImg;
        } else {
            return maxImg;
        }
    }

    void updateImage() {
        if (stage.isMaximized()) {
            imageView.setFitWidth(22);
            imageView.setFitHeight(22);
            imageView.setLayoutX((WIDTH - 22) / 2d);
            imageView.setLayoutY((HEIGHT - 22) / 2d);
            imageView.setScaleX(-1);
        } else {
            imageView.setFitWidth(20);
            imageView.setFitHeight(20);
            imageView.setLayoutX((WIDTH - 20) / 2d);
            imageView.setLayoutY((HEIGHT - 20) / 2d);
            imageView.setScaleX(-1);
        }
        if (clickHandler.isMouseEntered()) {
            imageView.setImage(currentImageGroup()[1]);
        } else {
            imageView.setImage(currentImageGroup()[0]);
        }
    }
}

File: src-ts/src/ui/stage/MaxResetButton.ts
import { Theme } from '../../theme/Theme.js';
import type { FXImage } from '../../javafx/ImageView.js';
import { WindowControlButton } from './WindowControlButton.js';
import { VStage } from './VStage.js';
import { VStageInitParams } from './VStageInitParams.js';

export class MaxResetButton extends WindowControlButton {
  private readonly maxImg: [FXImage, FXImage];
  private readonly rstImg: [FXImage, FXImage];
  private cornerRadiiCssStr = '';

  constructor(stage: VStage, initParams: VStageInitParams) {
    super(stage, initParams);
    this.maxImg = [
      Theme.current().windowMaximizeButtonNormalImage(),
      Theme.current().windowMaximizeButtonHoverImage(),
    ];
    this.rstImg = [
      Theme.current().windowResetWindowSizeButtonNormalImage(),
      Theme.current().windowResetWindowSizeButtonHoverImage(),
    ];
    this.updateImage();
  }

  protected init(initParams: VStageInitParams): void {
    if (initParams.iconifyButton) {
      this.cornerRadiiCssStr = '0';
    } else {
      this.cornerRadiiCssStr = '0 0 0 4px';
    }
  }

  protected onMouseEntered(): void {
    super.onMouseEntered();
    this.imageView.setImage(this.currentImageGroup()[1]);
  }

  protected onMouseExited(): void {
    super.onMouseExited();
    this.imageView.setImage(this.currentImageGroup()[0]);
  }

  protected onMouseClicked(): void {
    void this.stage.setMaximized(!this.stage.isMaximized());
  }

  protected getCornerRadiiCss(): string { return this.cornerRadiiCssStr; }

  protected getNormalImage(): FXImage {
    return Theme.current().windowMaximizeButtonNormalImage();
  }
  protected getHoverImage(): FXImage {
    return Theme.current().windowMaximizeButtonHoverImage();
  }

  private currentImageGroup(): [FXImage, FXImage] {
    return this.stage.isMaximized() ? this.rstImg : this.maxImg;
  }

  updateImage(): void {
    if (this.stage.isMaximized()) {
      this.imageView.setFitWidth(22);
      this.imageView.setFitHeight(22);
      this.imageView.setLayoutX((WindowControlButton.WIDTH - 22) / 2);
      this.imageView.setLayoutY((WindowControlButton.HEIGHT - 22) / 2);
      // Java: imageView.setScaleX(-1) — horizontal mirror only. The TS
      // port previously called setScale(-1), which scaled both axes and
      // relied on the source images being vertically symmetric; using
      // setScaleX preserves the exact Java semantic.
      this.imageView.setScaleX(-1);
    } else {
      this.imageView.setFitWidth(20);
      this.imageView.setFitHeight(20);
      this.imageView.setLayoutX((WindowControlButton.WIDTH - 20) / 2);
      this.imageView.setLayoutY((WindowControlButton.HEIGHT - 20) / 2);
      this.imageView.setScaleX(-1);
    }
    if (this.isMouseEntered()) {
      this.imageView.setImage(this.currentImageGroup()[1]);
    } else {
      this.imageView.setImage(this.currentImageGroup()[0]);
    }
  }
}
